addDependencyTracker.js

```diff
@@ -0,0 +1,58 @@
+function addDependencyTracker(plugin, enableInvalidation) {
+  if (plugin.prototype && plugin.prototype.transform) {
+    // we don't track dependencies for legacy plugins.
+    return plugin;
+  }
+  if (!enableInvalidation) {
+    // Dependency invalidation isn't enabled, so no need to track.
+    return plugin;
+  }
+  if (typeof plugin.getDependencies === "function") {
+    // Don't add the dependency tracker if it's already added
+    return plugin;
+  }
+  // This variable in our closure allows us to share dependencies from
+  // the ast plugin that we can't access with the ast plugin generator that
+  // we can access.
+  let lastDependencies = {};
+  let trackedPlugin = (env) => {
+    let realPlugin = plugin(env);
+    let visitor = realPlugin.visitor;
+    let origProgram = visitor.Program;
+    let origEnter, origExit, origKeys;
+    if (origProgram) {
+      if (typeof origProgram === "function") {
+        origEnter = origProgram;
+        origExit = undefined;
+        origKeys = undefined;
+      } else {
+        origEnter = origProgram.enter;
+        origExit = origProgram.exit;
+        origKeys = origProgram.keys;
+      }
+    }
+    visitor.Program = {
+      keys: origKeys,
+      enter: (node) => {
+        if (realPlugin.resetDependencies) {
```

Comment:
I think we only want to do this for the main Program (e.g. the whole file), `Program` itself exists once for the whole template/file but also once for every block statement. Eventually (when Ember updates to glimmer-vm 0.41+) we will be able to migrate to the `Template` hook here instead, but for now you need to track and only `resetDependencies` the first time `Program.enter` is called.

I think we should also add some tests that have templates with block component invocation (which presumably would have made `resetDependencies` get called too many times).

Reply:
Good catch. Tests added and code/tests fixed.